modules/client/src/internal/client/encoding.ts

```diff
@@ -164,26 +164,52 @@ export class ClientEncoding extends ClientCore implements IClientEncoding {
       args,
     ]);
     const pspAddress = this.web3.getAddress("pluginSetupProcessorAddress");
-    // Grant ROOT_PERMISION in the DAO to the PSP
-    const grantAction = this.grantAction(daoAddress, {
-      where: daoAddress,
+    
+    // Grant UPGRADE_PLUGIN_PERMISSION in the plugin to the PSP
+    const grantUpgradeAction = this.grantAction(daoAddress, {
+      where: params.pluginAddress,
       who: pspAddress,
-      permission: Permissions.ROOT_PERMISSION,
+      permission: Permissions.UPGRADE_PLUGIN_PERMISSION,
     });
-    // Revoke ROOT_PERMISION in the DAO to the PSP
-    const revokeAction = this.revokeAction(daoAddress, {
-      where: daoAddress,
+    // Revoke UPGRADE_PLUGIN_PERMISSION in the plugin to the PSP
+    const revokeUpgradeAction = this.revokeAction(daoAddress, {
+      where: params.pluginAddress,
       who: pspAddress,
-      permission: Permissions.ROOT_PERMISSION,
+      permission: Permissions.UPGRADE_PLUGIN_PERMISSION,
     });
+    // If the update requests permissions to be granted or revoked, the PSP needs temporary `ROOT_PERMISSION_ID` permission
+    if (params.permissions.length > 0) {
+      const grantRootAction = this.grantAction(daoAddress, {
+        where: daoAddress,
+        who: pspAddress,
+        permission: Permissions.ROOT_PERMISSION,
+      });
+      // Revoke ROOT_PERMISSION in the DAO to the PSP
+      const revokeRootAction = this.revokeAction(daoAddress, {
+        where: daoAddress,
+        who: pspAddress,
+        permission: Permissions.ROOT_PERMISSION,
+      });
+      return [
```

Comment:
We should rename the function, e.g., to `applyUpdateActionAndPermissionsBlock`. The function name suggests that you get a single action back, which is not the case. This change should also be reflected and explained in the jsdoc comment.
Descriptive/explicit naming is a must and we should never do shortcuts on this one.